feat(RingTheory/Polynomial/SmallDegreeVieta): polynomial versions of results in Algebra.QuadraticDiscriminant

mans0954 opened the pull request:
This PR continues the work from #25605.

Original PR: https://github.com/leanprover-community/mathlib4/pull/25605

mans0954 committed:
Polynomial versions of results in Algebra.QuadraticDiscriminant

mans0954 committed:
Fix name

mans0954 committed:
quadratic_roots_of_discrim_ne_sq

mans0954 committed:
Remove unused have

mans0954 committed:
Remove unused arg

mans0954 committed:
quadratic_roots_of_discrim_eq_sq

mans0954 committed:
Golf

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

mans0954 committed:
Factorize a quadratic polynomial with two roots

mans0954 committed:
WiP

mans0954 committed:
Factor

mans0954 committed:
Remove unused arg

mans0954 committed:
Refactor

mans0954 committed:
golf

mans0954 committed:
Golf more

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

mans0954 committed:
Reset Algebra/QuadraticDiscriminant to master

mans0954 committed:
Remove whitespace

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
No longer need C_2

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

mans0954 commented:
## Comments from Original PR #25605

*This section contains 1 comment(s) from the original PR, excluding bot comments.*

---

**@github-actions** ([2025-06-08 20:22 UTC](https://github.com/leanprover-community/mathlib4/pull/25605#issuecomment-2954260822)):
### PR summary [d6898ad6d1](https://github.com/leanprover-community/mathlib4/pull/25605/commits/3fc0772b9fa7bc21b50f0c1a3482c90e17c59cf8)

<details><summary>

#### Import changes for modified files

</summary>

Dependency changes

| File | Base Count | Head Count | Change |
| --- | --- | --- | --- |
| Mathlib.RingTheory.Polynomial.SmallDegreeVieta | 1233 | 1240 | +7 (+0.57%) |


<details><summary>Import changes for all files</summary>

|Files|Import difference|
|-|-|
| `Mathlib.RingTheory.Polynomial.SmallDegreeVieta`|7|

</details>

</details>

---

<details><summary>

#### Declarations diff

</summary>

`+ factorize_quadratic_of_discrim_eq_sq`
`+ quadratic_eq_zero_iff`
`+ quadratic_eq_zero_iff_of_discrim_eq_zero`
`+ quadratic_ne_zero`
`+ quadratic_ne_zero_of_discrim_ne_sq`
`+ quadratic_roots_iff_of_discrim_eq_sq`
`+ quadratic_roots_of_discrim_eq_sq`
`+ quadratic_roots_of_discrim_ne_sq`
<details>
  <summary>You can run this locally as follows</summary>


```bash
## summary with just the declaration names:
./scripts/declarations_diff.sh <optional_commit>

## more verbose report:
./scripts/declarations_diff.sh long <optional_commit>
```
</details>

The doc-module for `script/declarations_diff.sh` contains some details about this script.

</details>

---

<details><summary>No changes to technical debt.</summary>

You can run this locally as
```
./scripts/technical-debt-metrics.sh pr_summary
```
* The `relative` value is the weighted *sum* of the differences with weight given by the *inverse* of the current value of the statistic.
* The `absolute` value is the `relative` value divided by the total sum of the inverses of the current values (i.e. the weighted *average* of the differences).
</details>

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +101,73 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+variable {a b c : R}
+
+/-- Roots of a quadratic equation. -/
+theorem quadratic_eq_zero_iff [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
+    (h : discrim a b c = s * s) (x : R) :
+    IsRoot (a • X ^ 2 + b • X + C c) x ↔ x = (-b + s) / (2 * a) ∨ x = (-b - s) / (2 * a) := by
+  simp only [IsRoot.def, eval_add, eval_smul, eval_pow, eval_X, smul_eq_mul, eval_C]
+  rw [sq, _root_.quadratic_eq_zero_iff ha h]
```
```suggestion
theorem isRoot_quadratic_iff [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
    (h : discrim a b c = s * s) (x : R) :
    IsRoot (a • X ^ 2 + b • X + C c) x ↔ x = (-b + s) / (2 * a) ∨ x = (-b - s) / (2 * a) := by
  rw [← quadratic_eq_zero_iff ha h]; simp [pow_two]
```

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +101,73 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+  rw [sq, _root_.quadratic_eq_zero_iff ha h]
+
+/-- Root of a quadratic when its discriminant equals zero -/
+theorem quadratic_eq_zero_iff_of_discrim_eq_zero [NeZero (2 : R)] (ha : a ≠ 0)
+    (h : discrim a b c = 0) (x : R) : IsRoot (a • X ^ 2 + b • X + C c) x ↔ x = -b / (2 * a) := by
+  simp only [IsRoot.def, eval_add, eval_smul, eval_pow, eval_X, smul_eq_mul, eval_C]
+  rw [sq, _root_.quadratic_eq_zero_iff_of_discrim_eq_zero ha h]
```
```suggestion
theorem isRoot_quadratic_iff_of_discrim_eq_zero [NeZero (2 : R)] (ha : a ≠ 0)
    (h : discrim a b c = 0) (x : R) :
    IsRoot (a • X ^ 2 + b • X + C c) x ↔ x = -b / (2 * a) := by
  rw [← quadratic_eq_zero_iff_of_discrim_eq_zero ha h]; simp [pow_two]
```

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +101,73 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+  rw [sq, _root_.quadratic_eq_zero_iff_of_discrim_eq_zero ha h]
+
+/-- A quadratic has no root if its discriminant has no square root. -/
+theorem quadratic_ne_zero_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) (x : R) :
+    ¬ IsRoot (a • X ^ 2 + b • X + C c) x := by
+  simp only [IsRoot.def, eval_add, eval_smul, eval_pow, eval_X, smul_eq_mul, eval_C]
+  rw [← ne_eq, sq]
+  exact _root_.quadratic_ne_zero_of_discrim_ne_sq h _
```
```suggestion
theorem not_isRoot_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) (x : R) :
    ¬ IsRoot (a • X ^ 2 + b • X + C c) x := by
  convert quadratic_ne_zero_of_discrim_ne_sq h x using 1; simp [pow_two]
```

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +101,73 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+  rw [← ne_eq, sq]
+  exact _root_.quadratic_ne_zero_of_discrim_ne_sq h _
+
+lemma quadratic_ne_zero (ha : a ≠ 0) : (a • X ^ 2 + b • X + C c) ≠ 0 := by
+  have hc : (a • X ^ 2 + b • X + C c).coeff 2 = a := by
+    simp [coeff_add, coeff_C_mul, coeff_smul, coeff_X_of_ne_one (Nat.add_one_add_one_ne_one),
+      coeff_C_ne_zero (n:=2) ((Nat.zero_ne_add_one 1).symm)]
+  rw [← hc] at ha
+  by_contra hx
+  exact ha (congrFun (congrArg coeff hx) 2)
```
```suggestion
lemma quadratic_ne_zero (ha : a ≠ 0) : a • X ^ 2 + b • X + C c ≠ 0 :=
  fun hx ↦ ha (by rw [show a = (a • X ^ 2 + b • X + C c).coeff 2 by simp [coeff_X], hx, coeff_zero])
```

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +101,73 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+  by_contra hx
+  exact ha (congrFun (congrArg coeff hx) 2)
+
+theorem quadratic_roots_iff_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0)
+    {z s : R} (h : discrim a b c = s * s) :
+    z ∈ (a • X ^ 2 + b • X + C c).roots ↔ z = (-b + s) / (2 * a) ∨ z = (-b - s) / (2 * a) := by
+  rw [mem_roots (quadratic_ne_zero ha), quadratic_eq_zero_iff ha h]
```
```suggestion
theorem mem_roots_quadratic_iff_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0)
    {z s : R} (h : discrim a b c = s * s) :
    z ∈ (a • X ^ 2 + b • X + C c).roots ↔ z = (-b + s) / (2 * a) ∨ z = (-b - s) / (2 * a) := by
  rw [mem_roots (quadratic_ne_zero ha), isRoot_quadratic_iff ha h]
```

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +101,73 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+    z ∈ (a • X ^ 2 + b • X + C c).roots ↔ z = (-b + s) / (2 * a) ∨ z = (-b - s) / (2 * a) := by
+  rw [mem_roots (quadratic_ne_zero ha), quadratic_eq_zero_iff ha h]
+
+theorem factorize_quadratic_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
+      (h : discrim a b c = s * s) :
+    a • X ^ 2 + b • X + C c = a • (X - C ((-b + s) / (2 * a))) * (X - C ((-b - s) / (2 * a))) := by
```
```suggestion
theorem quadratic_eq_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
    (h : discrim a b c = s * s) :
    a • X ^ 2 + b • X + C c = a • (X - C ((-b + s) / (2 * a))) * (X - C ((-b - s) / (2 * a))) := by
```

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +101,73 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+  · field_simp
+  · field_simp; simp [pow_two, ← h, discrim]; ring
+
+theorem quadratic_roots_of_discrim_ne_sq (ha : a ≠ 0) (h : ∀ s : R, discrim a b c ≠ s^2) :
+    (a • X ^ 2 + b • X + C c).roots = ∅ := Multiset.eq_zero_of_forall_notMem (fun r => by
+  by_contra hc
+  exact (quadratic_ne_zero_of_discrim_ne_sq h _) ((mem_roots (quadratic_ne_zero ha)).mp hc))
```
```suggestion
theorem roots_quadratic_of_discrim_ne_sq (ha : a ≠ 0) (h : ∀ s : R, discrim a b c ≠ s^2) :
    (a • X ^ 2 + b • X + C c).roots = ∅ :=
  Multiset.eq_zero_of_forall_notMem fun r hc => not_isRoot_of_discrim_ne_sq h r
    ((mem_roots (quadratic_ne_zero ha)).mp hc)
```

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +101,73 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+  by_contra hc
+  exact (quadratic_ne_zero_of_discrim_ne_sq h _) ((mem_roots (quadratic_ne_zero ha)).mp hc))
+
+theorem quadratic_roots_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
+    (h : discrim a b c = s * s) :
+    (a • X ^ 2 + b • X + C c).roots = {(-b + s) / (2 * a), (-b - s) / (2 * a)} := by
+  rw [factorize_quadratic_of_discrim_eq_sq ha h, Polynomial.roots_mul
+    (by rw [← factorize_quadratic_of_discrim_eq_sq ha h]; exact quadratic_ne_zero ha)]
+  simp_all only [ne_eq, Algebra.smul_mul_assoc, not_false_eq_true, roots_smul_nonzero,
+    roots_X_sub_C, Multiset.singleton_add, Multiset.insert_eq_cons]
```
```suggestion
theorem roots_quadratic_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
    (h : discrim a b c = s * s) :
    (a • X ^ 2 + b • X + C c).roots = {(-b + s) / (2 * a), (-b - s) / (2 * a)} := by
  rw [quadratic_eq_of_discrim_eq_sq ha h, Polynomial.roots_mul
      (quadratic_eq_of_discrim_eq_sq ha h ▸ quadratic_ne_zero ha),
    roots_smul_nonzero _ ha, roots_X_sub_C, roots_X_sub_C]
  rfl
```

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

kckennylau commented:
After rereading the whole file, I found that there might be some refactoring that might be beneficial.

The following lemma `roots_quadratic_eq_pair_iff_of_ne_zero`:
```lean
lemma roots_quadratic_eq_pair_iff_of_ne_zero [CommRing R] [IsDomain R] {a b c x1 x2 : R} (ha : a ≠ 0) :
    (C a * X ^ 2 + C b * X + C c).roots = {x1, x2} ↔ b = -a * (x1 + x2) ∧ c = a * x1 * x2 :=
```
seems to be a useful intermediate step for `roots_quadratic_of_discrim_eq_sq` (your last theorem);

and the proof of `roots_quadratic_eq_pair_iff_of_ne_zero` has the following intermediate step:
```lean
have roots_of_ne_zero_of_vieta (hvieta : b = -a * (x1 + x2) ∧ c = a * x1 * x2) :
      (C a * X ^ 2 + C b * X + C c).roots = {x1, x2} := by
```
which could be extracted out as a lemma.

and the proof of this `roots_of_ne_zero_of_vieta` involves an explicit equality of polynomials:
```lean
C a * X ^ 2 + C b * X + C c = C a * (X - C x1) * (X - C x2)
```
which seems useful for proving your `quadratic_eq_of_discrim_eq_sq`, so I think this step should also be factored out.

You could try it yourself to become more familiar, or I could also do it for you, which one would you like?

kckennylau commented:
As a separate comment, it seems like you want to treat the case `discrim_eq_zero` separately as a special case. Do you think it would be a good idea to do the case `discrim_eq_zero` for the rest of the file?

kckennylau commented:
And finally, as another separate comment, it seems like `C a * p` is preferred over `a • p`, but I would ask in Zulip to make sure first. The evidence I have is just the direction of the lemma `smul_eq_C_mul`. It's an unfortunate situation that both notations seem to exist.

mans0954 committed:
Convert to C a * X format

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

mans0954 committed:
Use roots_quadratic_eq_pair_iff_of_ne_zero in theorem roots_quadratic_of_discrim_eq_sq

mans0954 committed:
Extract lemmas

mans0954 committed:
use quadratic_eq_of_vieta

mans0954 committed:
fix

mans0954 commented:
@kckennylau I've extracted `roots_of_ne_zero_of_vieta` - was there somewhere you intended it to be used? It seems to be one implication of `roots_quadratic_eq_pair_iff_of_ne_zero` - which I've now used in `roots_quadratic_of_discrim_eq_sq`.

I've extracted out the intermediate step as `quadratic_eq_of_vieta` which then allows me to simplify the proof of `quadratic_eq_of_discrim_eq_sq`.

I've also switched to `C a * X` as that seems to be used almost everywhere else.

mans0954 committed:
Fix

mans0954 committed:
Reduce hypothesis

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
roots_of_ne_zero_of_vieta

mans0954 committed:
Golf

mans0954 committed:
Sketch solution

mans0954 commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +108,63 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+theorem roots_quadratic_of_discrim_ne_sq (ha : a ≠ 0) (h : ∀ s : R, discrim a b c ≠ s^2) :
+    (C a * X ^ 2 + C b * X + C c).roots = ∅ :=
+  Multiset.eq_zero_of_forall_notMem fun r hc => not_isRoot_of_discrim_ne_sq h r
+    ((mem_roots (quadratic_ne_zero ha)).mp hc)
```
I've quickly sketched out a solution to this. I expect it can be golfed.

kckennylau replied:
Yeah, you can prove it by using cases on whether a = 0 (basically just using the original proof).

kckennylau commented:
> @kckennylau I've extracted `roots_of_ne_zero_of_vieta` - was there somewhere you intended it to be used? It seems to be one implication of `roots_quadratic_eq_pair_iff_of_ne_zero` - which I've now used in `roots_quadratic_of_discrim_eq_sq`.

I see. I think I intended it to be used in exactly that lemma. If you just use the reverse implication of the iff, why not use `roots_of_ne_zero_of_vieta`?

mans0954 committed:
roots_quadratic_of_discrim_ne_sq rcases

mans0954 committed:
roots_quadratic_of_discrim_ne_sq aesop

mans0954 committed:
roots_of_ne_zero_of_vieta

mans0954 commented:
> I see. I think I intended it to be used in exactly that lemma. If you just use the reverse implication of the iff, why not use `roots_of_ne_zero_of_vieta`?

Okay, done.

kckennylau commented:
@mans0954 wow, you found an even shorter proof than me! I had to do some debugging to see how the proof works, and I'm amazed. It actually derives that the polynomial is not zero from the fact that `r` is in the `roots`; because `roots` is defined to only be nonempty when the polynomial is not zero!

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +111,58 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+      C a * (X - C ((-b + s) / (2 * a))) * (X - C ((-b - s) / (2 * a))) :=
+  quadratic_eq_of_vieta (vieta_of_discrim_eq_sq ha h)
+
+theorem roots_quadratic_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) :
+    (C a * X ^ 2 + C b * X + C c).roots = ∅ :=
+  Multiset.eq_zero_of_forall_notMem (fun r hc => not_isRoot_of_discrim_ne_sq h r
+      ((mem_roots (by aesop)).mp hc))
```
```suggestion
theorem roots_quadratic_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) :
    (C a * X ^ 2 + C b * X + C c).roots = ∅ :=
  Multiset.eq_zero_of_forall_notMem (fun r hc => not_isRoot_of_discrim_ne_sq h r
    (mem_roots'.mp hc).2)
```

kckennylau commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +111,58 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+theorem roots_quadratic_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
+    (h : discrim a b c = s * s) :
+    (C a * X ^ 2 + C b * X + C c).roots = {(-b + s) / (2 * a), (-b - s) / (2 * a)} :=
+  (roots_of_ne_zero_of_vieta ha) (vieta_of_discrim_eq_sq ha h)
```
```suggestion
  roots_of_ne_zero_of_vieta ha (vieta_of_discrim_eq_sq ha h)
```

kckennylau reviewed (approved):
LGTM!

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

mans0954 committed:
Remove unused simp

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

ocfnash commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -15,6 +16,10 @@ namespace Polynomial
 
 variable {R T S : Type*}
 
+lemma quadratic_ne_zero [Semiring R] {a b c : R} (ha : a ≠ 0) : C a * X ^ 2 + C b * X + C c ≠ 0 :=
+  fun hx ↦ ha (by rw [show a = (C a * X ^ 2 + C b * X + C c).coeff 2 by
+    simp, hx, coeff_zero])
```
I don't think this deserves to be a lemma. It is only a fragment of the contrapositive of the following useful lemma:
```suggestion
@[simp]
lemma quadratic_eq_zero_iff [Semiring R] {a b c : R} :
    C a * X ^ 2 + C b * X + C c = 0 ↔ a = 0 ∧ b = 0 ∧ c = 0 := by
  constructor
  · rw [Polynomial.ext_iff]
    exact fun h ↦ ⟨by simpa using h 2, by simpa using h 1, by simpa using h 0⟩
  · rintro ⟨rfl, rfl, rfl⟩
    simp
```
However what would be really desirable here is a `simproc` which worked for any degree.

mans0954 replied:
How about:
```
lemma zero_iff [Semiring R] {p : R[X]} : p = 0 ↔ ∀ (n : ℕ), coeff p n = 0 := ext_iff

lemma not_zero_iff [Semiring R] {p : R[X]} : ¬ p = 0 ↔ ∃ (n : ℕ), ¬ coeff p n = 0 := by
  rw [zero_iff, not_forall]
```
which can then be invoked with `not_zero_iff.mpr ⟨2, by simp [ha]⟩` to get the desired goal.

ocfnash commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -56,20 +61,26 @@ lemma eq_mul_mul_of_aroots_quadratic_eq_pair [CommRing T] [CommRing S] [IsDomain
   exact eq_mul_mul_of_roots_quadratic_eq_pair haroots
 
+lemma quadratic_eq_of_vieta [CommRing R] {a b c x1 x2 : R}
+    (hvieta : b = -a * (x1 + x2) ∧ c = a * x1 * x2) :
```
We do not package assumptions together using conjunction like this as it is less extensible than making them separate assumptions:
```suggestion
    (hsum : b = -a * (x1 + x2)) (hprod : c = a * x1 * x2) :
```
(similarly below)

ocfnash commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +111,58 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+      C a * (X - C ((-b + s) / (2 * a))) * (X - C ((-b - s) / (2 * a))) :=
+  quadratic_eq_of_vieta (vieta_of_discrim_eq_sq ha h)
+
+theorem roots_quadratic_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) :
```
You're using a mixture of `s * s` and `s ^ 2`: this will make for a very annoying API. I think either is fine but we must be consistent.

mans0954 replied:
Okay, let's use `s ^ 2`.

ocfnash commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +111,58 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+    IsRoot (C a * X ^ 2 + C b * X + C c) x ↔ x = -b / (2 * a) := by
+  rw [← quadratic_eq_zero_iff_of_discrim_eq_zero ha h]; simp [pow_two]
+
+/-- A quadratic has no root if its discriminant has no square root. -/
+theorem not_isRoot_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) (x : R) :
+    ¬ IsRoot (C a * X ^ 2 + C b * X + C c) x := by
```
I think this is half of a much more useful result:
```lean
/-- Quantifier elimination for a quadratic equation in characteristic `≠ 2`. -/
lemma Polynomial.exists_isRoot_quadratic_iff [NeZero (2 : R)] (ha : a ≠ 0) :
    (∃ x, IsRoot (C a * X ^ 2 + C b * X + C c) x) ↔ ∃ s, discrim a b c = s ^ 2 := by
```
I admit this means that people with `(2 : R) = 0` or `a = 0` are no longer served by this lemma but I still think this is the right API.

mans0954 replied:
We could deduce `exists_isRoot_quadratic_iff` from `not_isRoot_of_discrim_ne_sq` and `isRoot_quadratic_iff`?

ocfnash commented on Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean:
```diff
@@ -100,4 +111,58 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
+      C a * (X - C ((-b + s) / (2 * a))) * (X - C ((-b - s) / (2 * a))) :=
+  quadratic_eq_of_vieta (vieta_of_discrim_eq_sq ha h)
+
+theorem roots_quadratic_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) :
+    (C a * X ^ 2 + C b * X + C c).roots = ∅ :=
+  Multiset.eq_zero_of_forall_notMem (fun r hc => not_isRoot_of_discrim_ne_sq h r
+    (mem_roots'.mp hc).2)
```
This lemma is basically just `Multiset.eq_zero_of_forall_notMem`. I'm not convinced we want it.

ocfnash reviewed (commented):
In addition to my code comments I have two more general remarks:

1. The naming is not in line with our style guide: lemmas are supposed to be named based on the type of their hypotheses and conclusions. It is only definitions that use informal names. I note that this file unfortunately already breaks this rule by using `quadratic` but I don't think we should double down on this with the `vieta` naming
2. Is there a reason to prefer the spelling `C a * X ^ 2 + C b * X + C c` vs `a • X ^ 2 + b • X + C c`?

mans0954 committed:
not_zero_iff

mans0954 committed:
Update Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

mans0954 committed:
hsum and hprod

mans0954 committed:
standardise on s ^ 2

mans0954 committed:
exists_isRoot_quadratic_iff

mans0954 committed:
Remove roots_quadratic_of_discrim_ne_sq

mans0954 commented:
> 1. The naming is not in line with our style guide: lemmas are supposed to be named based on the type of their hypotheses and conclusions. It is only definitions that use informal names. I note that this file unfortunately already breaks this rule by using `quadratic` but I don't think we should double down on this with the `vieta` naming

I always seem to fail to interpret this correctly in practice.

Instead of `quadratic_eq_of_vieta` should it be `polynomial_eq_of_degree_le_two_neg_mul_sum_mul_mul` or something like that?

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

ocfnash commented:
I regret that I continue to fail to find time to review this so I am unassigning myself.

chrisflav commented:
Since very recently we have [Polynomial.disc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Resultant/Basic.html#Polynomial.disc) and I personally think we should stop using [discrim](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/QuadraticDiscriminant.html#discrim) in favor of the more general `Polynomial.disc`. In particular this would mean changing the statements in this PR to be about `p : R[X]` with `hp : p.natDegree = 2` instead of explicit expressions like `C a * X ^ 2 + C b * X + C c`.
I have opened https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/discriminants.20of.20low.20degree.20polynomials/with/538010519 to discuss this.

mans0954 committed:
Merge branch 'master' into mans0954/quadratic-polynomials

mans0954 committed:
Get it building again
